### PR TITLE
Policy combiner

### PR DESCRIPTION
**Pull request by tvmarino**
Combines two tf-agents policies with the given signature spec in get_input_signature.

### CODE REVIEW

File: compiler_opt/tools/combine_tfa_policies_lib.py [hunk 1]
```diff
@@ -0,0 +1,193 @@
...
+
+    distribution = tf.cond(
+        tf.math.reduce_all(tf.equal(switch_tensor, self.high_low_tensor)), f0,
+        f1)
+    return tf_agents.trajectories.PolicyStep(
+        action=self._create_distribution(distribution), state=policy_state)
+
+
+@gin.configurable()
+def get_input_signature():
```
Comment (mtrofin) [inline compiler_opt/tools/combine_tfa_policies_lib.py:139]: this should come in as a parameter from the user, ie, in this case, something the tool passes in.

also, just use the config registry, like here: https://github.com/google/ml-compiler-opt/blob/dc45cf283a412fa596b9627382a0aaadf059881c/compiler_opt/tools/generate_test_model.py#L46

File: compiler_opt/tools/combine_tfa_policies_lib.py [hunk 2]
```diff
@@ -0,0 +1,193 @@
...
+                ])
+  inputs.update({
+      "model_selector":
+          tf.TensorSpec(shape=(2,), dtype=tf.uint64, name="model_selector")
+  })
+  return ts.time_step_spec(inputs)
+
+
+@gin.configurable()
+def get_action_spec():
```
Comment (mtrofin) [inline compiler_opt/tools/combine_tfa_policies_lib.py:191]: same idea as above - this is something the ProblemConfiguration should be able to provide.

File: compiler_opt/tools/combine_tfa_policies.py [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+"""Runs the policy combiner."""
+from absl import app
+
+import tensorflow as tf
+
+from compiler_opt.rl import policy_saver
+from compiler_opt.tools import combine_tfa_policies_lib as cfa_lib
+
+
+def main(_):
```
Comment (mtrofin) [inline compiler_opt/tools/combine_tfa_policies.py:24]: use absl flags to get the 2 policy paths in as command line flags. I'd suggest using one multi-string flag (`policies` or like that).

you also need a flag for the output policy path.

File: compiler_opt/tools/combine_tfa_policies_lib_test.py [hunk 4]
```diff
@@ -0,0 +1,149 @@
...
+from tf_agents.trajectories import time_step as ts
+import tf_agents
+from tf_agents.specs import tensor_spec
+from tf_agents.trajectories import policy_step
+import hashlib
+import numpy as np
+
+
+class AddOnePolicy(tf_agents.policies.TFPolicy):
+  """Test policy which adds one to obs feature."""
```
Comment (mtrofin) [inline compiler_opt/tools/combine_tfa_policies_lib_test.py:30]: nit: s/one/1 (or just say "increments") - easier to read (otherwise may think "one what", rather than realize it's the number `1`)

File: compiler_opt/tools/combine_tfa_policies_lib_test.py [hunk 5]
```diff
@@ -0,0 +1,149 @@
...
+    return ()
+
+  def _action(self, time_step, policy_state, seed):
+    observation = time_step.observation['obs'][0]
+    action = tf.reshape(observation + 1, (1,))
+    return policy_step.PolicyStep(action, policy_state)
+
+
+class SubtractOnePolicy(tf_agents.policies.TFPolicy):
+  """Test policy which subtracts one to obs feature."""
```
Comment (mtrofin) [inline compiler_opt/tools/combine_tfa_policies_lib_test.py:53]: same nit as above.

File: compiler_opt/tools/combine_tfa_policies_lib_test.py [hunk 6]
```diff
@@ -0,0 +1,149 @@
...
+from compiler_opt.tools import combine_tfa_policies_lib
+from tf_agents.trajectories import time_step as ts
+import tf_agents
+from tf_agents.specs import tensor_spec
+from tf_agents.trajectories import policy_step
+import hashlib
+import numpy as np
+
+
+class AddOnePolicy(tf_agents.policies.TFPolicy):
```
Comment (mtrofin) [inline compiler_opt/tools/combine_tfa_policies_lib_test.py:29]: can you add a comment that explains that the _distribution, _variables, and _action members are boylerplate coming from tf_agents? (that's... what it is, right?)

File: compiler_opt/tools/combine_tfa_policies_lib_test.py [hunk 7]
```diff
@@ -0,0 +1,149 @@
...
+    policy2 = SubtractOnePolicy()
+    combined_policy = combine_tfa_policies_lib.CombinedTFPolicy(
+        tf_policies={
+            'add_one': policy1,
+            'subtract_one': policy2
+        },
+        time_step_spec=observation_spec,
+        action_spec=action_spec)
+
+    m = hashlib.md5()
```
Comment (mtrofin) [inline compiler_opt/tools/combine_tfa_policies_lib_test.py:98]: nit: you could refactor the calculation here and reuse it between the 2 tests. if you give the function a name like "client_side_model_selector_calculation" or something like that, a reader will "get it" that this is part of the contract with the user.

File: compiler_opt/tools/combine_tfa_policies_lib.py [hunk 8]
```diff
@@ -0,0 +1,193 @@
...
+      high_low_tensors.append(
+          tf.stack([
+              tf.constant(
+                  int.from_bytes(m.digest()[8:], "little"), dtype=tf.uint64),
+              tf.constant(
+                  int.from_bytes(m.digest()[:8], "little"), dtype=tf.uint64)
+          ]))
+    self.high_low_tensors = tf.stack(high_low_tensors)
+
+    m = hashlib.md5()
```
Comment (mtrofin) [inline compiler_opt/tools/combine_tfa_policies_lib.py:59]: you could add a comment here referencing the related LLVM side commit: https://github.com/llvm/llvm-project/pull/96276

File: compiler_opt/tools/combine_tfa_policies.py [hunk 9]
```diff
@@ -0,0 +1,82 @@
...
+from compiler_opt.rl import policy_saver
+from compiler_opt.rl import registry
+from compiler_opt.tools import combine_tfa_policies_lib as cfa_lib
+
+_COMBINE_POLICIES_NAMES = flags.DEFINE_multi_string(
+    'policies_names', [],
+    'List in order of policy names for combined policies.')
+_COMBINE_POLICIES_PATHS = flags.DEFINE_multi_string(
+    'policies_paths', [],
+    'List in order of policy paths for combined policies.')
```
Comment (mtrofin) [inline compiler_opt/tools/combine_tfa_policies.py:32]: nit: maybe say here and with the policy names that the orders must match? well.. either that or you could do like a trivial parser, like "name:/policy/path" and just split on `:`. Up to you.

File: compiler_opt/tools/combine_tfa_policies.py [hunk 10]
```diff
@@ -0,0 +1,82 @@
...
+_GIN_BINDINGS = flags.DEFINE_multi_string(
+    'gin_bindings', [],
+    'Gin bindings to override the values set in the config files.')
+
+
+def main(_):
+  flags.mark_flag_as_required('policies_names')
+  flags.mark_flag_as_required('policies_paths')
+  flags.mark_flag_as_required('combined_policy_path')
+  assert len(_COMBINE_POLICIES_NAMES.value) == len(
```
Comment (mtrofin) [inline compiler_opt/tools/combine_tfa_policies.py:46]: Do you mean to fail here or just have the assert as a readability tool? We ended up not doing `assert` for errors, and rather either throwing ValueError or - and I think it'd be more appropriate here - check and print the error and exit(1).

May be worth spelling out also in the input policy flag docs that currently we support exactly 2.

File: compiler_opt/tools/combine_tfa_policies.py [hunk 11]
```diff
@@ -0,0 +1,82 @@
...
+      _COMBINE_POLICIES_PATHS.value)
+  gin.add_config_file_search_path(
```
Comment (mtrofin) [inline compiler_opt/tools/combine_tfa_policies.py:48]: why?

File: compiler_opt/tools/combine_tfa_policies_lib.py [hunk 12]
```diff
@@ -0,0 +1,138 @@
...
+from tf_agents.typing import types
+from tf_agents.trajectories import policy_step
+import tensorflow_probability as tfp
+
+
+class CombinedTFPolicy(tf_agents.policies.TFPolicy):
+  """Policy which combines two target policies."""
+
+  def __init__(self, *args, tf_policies: Dict[str, tf_agents.policies.TFPolicy],
+               **kwargs):
```
Comment (mtrofin) [inline compiler_opt/tools/combine_tfa_policies_lib.py:32]: do we need the `kwargs` right now? let's not have it if not (we can always add later)

File: compiler_opt/tools/combine_tfa_policies_lib.py [hunk 13]
```diff
@@ -0,0 +1,138 @@
...
+    self.high_low_tensor = tf.constant([self.high, self.low], dtype=tf.uint64)
+
+  def _process_observation(
+      self, observation: types.NestedSpecTensorOrArray
+  ) -> Tuple[types.NestedSpecTensorOrArray, types.TensorOrArray]:
+    assert "model_selector" in self.sorted_keys
+    high_low_tensor = self.high_low_tensor
+    for name in self.sorted_keys:
+      if name in ["model_selector"]:
+        switch_tensor = observation.pop(name)[0]
```
Comment (mtrofin) [inline compiler_opt/tools/combine_tfa_policies_lib.py:70]: can you add a comment why element 0 is what we need here? (wasn't immediately obvious to me)

File: compiler_opt/tools/combine_tfa_policies_lib.py [hunk 14]
```diff
@@ -0,0 +1,138 @@
...
+            tf.equal(
+                tf.reduce_any(
+                    tf.reduce_all(
+                        tf.equal(high_low_tensor, self.high_low_tensors),
+                        axis=1)), True),
+            [high_low_tensor, self.high_low_tensors])
+
+    return observation, high_low_tensor
+
+  def _create_distribution(self, inlining_prediction):
```
Comment (mtrofin) [inline compiler_opt/tools/combine_tfa_policies_lib.py:83]: open q: can we document this isn't introducing a new behavior? or is there some way to delegate the implementation? (OK to document and add a TODO if no easy way to delegate. If you go the TODO path, please open an issue and tag it here, i.e. TODO(1234)..)

File: compiler_opt/tools/combine_tfa_policies_lib.py [hunk 15]
```diff
@@ -0,0 +1,138 @@
...
+              seed: Optional[types.Seed] = None) -> policy_step.PolicyStep:
+    new_observation = time_step.observation
+    new_observation, switch_tensor = self._process_observation(new_observation)
+    updated_step = ts.TimeStep(
+        step_type=time_step.step_type,
+        reward=time_step.reward,
+        discount=time_step.discount,
+        observation=new_observation)
+
+    def f0():
```
Comment (mtrofin) [inline compiler_opt/tools/combine_tfa_policies_lib.py:100]: nit: add a comment explaining this is because we currently only support 2 policies, and tag a TODO with an issue about it.

File: compiler_opt/tools/combine_tfa_policies_lib.py [hunk 16]
```diff
@@ -0,0 +1,138 @@
...
+      policy_state: types.NestedTensorSpec) -> policy_step.PolicyStep:
+    new_observation = time_step.observation
+    new_observation, switch_tensor = self._process_observation(new_observation)
+    updated_step = ts.TimeStep(
+        step_type=time_step.step_type,
+        reward=time_step.reward,
+        discount=time_step.discount,
+        observation=new_observation)
+
+    def f0():
```
Comment (mtrofin) [inline compiler_opt/tools/combine_tfa_policies_lib.py:124]: copy'n'paste of the above observation (current limit of 2... issue... comment...)

File: compiler_opt/tools/combine_tfa_policies_lib_test.py [hunk 17]
```diff
@@ -0,0 +1,151 @@
...
+            'add_one': policy1,
+            'subtract_one': policy2
+        },
+        time_step_spec=observation_spec,
+        action_spec=action_spec)
+
+    model_selector = client_side_model_selector_calculation('add_one')
+
+    state = ts.TimeStep(
+        discount=tf.constant(np.array([0.]), dtype=tf.float32),
```
Comment (mtrofin) [inline compiler_opt/tools/combine_tfa_policies_lib_test.py:117]: nit: can you use non-zero values for the input to the test, use like `42` :). 

0 could mask what happens if you have (1) zero-initialized values that (2) don't actually get passed to the underlying model.

File: compiler_opt/tools/combine_tfa_policies.py [hunk 18]
```diff
@@ -0,0 +1,82 @@
...
+  gin.parse_config_files_and_bindings(
+      _GIN_FILES.value, bindings=_GIN_BINDINGS.value, skip_unknown=False)
+
+  problem_config = registry.get_configuration()
+  expected_signature, action_spec = problem_config.get_signature_spec()
+  expected_signature.observation.update({
+      'model_selector':
+          tf.TensorSpec(shape=(2,), dtype=tf.uint64, name='model_selector')
+  })
+  assert len(_COMBINE_POLICIES_NAMES.value
```
Comment (mtrofin) [inline compiler_opt/tools/combine_tfa_policies.py:59]: Could you open an issue to support more than 2 policies, and add a comment here with a TODO(that issue number) that we should at a point support more than 2. That way (1) we don't forget and (2) a contributor could understand what the goal is and help accordingly (otherwise they may wonder "so why 2...")

**Review by mtrofin — COMMENTED** [review]
this looks good, just mostly nits

File: compiler_opt/tools/combine_tfa_policies_lib_test.py (hunk 17 above)
Comment (tvmarino) [reply compiler_opt/tools/combine_tfa_policies_lib_test.py:117]: Changed the obs to 42, the other parts of the TimeStep are not used for this test.

File: compiler_opt/tools/combine_tfa_policies_lib.py (hunk 12 above)
Comment (tvmarino) [reply compiler_opt/tools/combine_tfa_policies_lib.py:32]: I am just using kwargs and args to pass these to the TFPolicy constructor so if they are needed they would be needed by TFPolicy. I am not completely sure on that last part though.

File: compiler_opt/tools/combine_tfa_policies_lib.py (hunk 14 above)
Comment (tvmarino) [reply compiler_opt/tools/combine_tfa_policies_lib.py:83]: This could introduce new behavior, e.g., when combining a deterministic policy such as the saved ppo models. Such policies do not return a proper distribution, however, this is not a good behavior IMO as we use distributions in our trajectory collecting code. In general (my) preferred behavior is to always return a discrete distribution over the action space even if that distribution is supported only on a single action, i.e., deterministic distributions will always look like [1, 0] or [0, 1]. Note that this will only create a new behavior when calling distribution. The behavior of action is remains unchanged.

File: compiler_opt/tools/combine_tfa_policies.py (hunk 11 above)
Comment (tvmarino) [reply compiler_opt/tools/combine_tfa_policies.py:48]: I could not get the binary to discover common.gin without adding this line. If there is a different way to do this let me know.

File: compiler_opt/tools/combine_tfa_policies_lib.py (hunk 12 above)
Comment (mtrofin) [reply compiler_opt/tools/combine_tfa_policies_lib.py:32]: I understand that, my point is that we don't currently make use of it. So let's remove them, and if we ever need to add them back, that's easy to do.

File: compiler_opt/tools/combine_tfa_policies_lib.py (hunk 14 above)
Comment (mtrofin) [reply compiler_opt/tools/combine_tfa_policies_lib.py:83]: > Note that this will only create a new behavior when calling distribution. The behavior of action is remains unchanged.

This means `_create_distribution` isn't used for inference? Can we raise an exception then or something that says "not implemented, and intentionally so"?

File: compiler_opt/tools/combine_tfa_policies.py (hunk 11 above)
Comment (mtrofin) [reply compiler_opt/tools/combine_tfa_policies.py:48]: try adding common.gin to the list of gin files you pass in.

File: compiler_opt/tools/combine_tfa_policies.py (hunk 18 above)
Comment (boomanaiden154) [reply compiler_opt/tools/combine_tfa_policies.py:59]: It seems like we're still missing a TODO here?

File: compiler_opt/tools/combine_tfa_policies_lib.py (hunk 16 above)
Comment (boomanaiden154) [reply compiler_opt/tools/combine_tfa_policies_lib.py:124]: Looks like the TODO should be linked here too?

File: compiler_opt/tools/combine_tfa_policies_lib_test.py [hunk 19]
```diff
@@ -0,0 +1,151 @@
...
+    'obs':
+        tf.TensorSpec(dtype=tf.int32, shape=(), name='obs'),
+    'model_selector':
+        tf.TensorSpec(shape=(2,), dtype=tf.uint64, name='model_selector')
+})
+
+action_spec = tensor_spec.TensorSpec(shape=(1,), dtype=tf.int64)
+
+
+class FeatureImportanceTest(absltest.TestCase):
```
Comment (boomanaiden154) [inline compiler_opt/tools/combine_tfa_policies_lib_test.py:101]: This class name doesn't seem particularly relevant and should probably be switched?

**Review by boomanaiden154 — COMMENTED** [review]
Just nits from my end.

File: compiler_opt/tools/combine_tfa_policies.py (hunk 11 above)
Comment (tvmarino) [reply compiler_opt/tools/combine_tfa_policies.py:48]: I tried passing it as part of gin_files but that didn't work, that's why I added it. I can remove it if you want.

File: compiler_opt/tools/combine_tfa_policies_lib.py (hunk 14 above)
Comment (tvmarino) [reply compiler_opt/tools/combine_tfa_policies_lib.py:83]: Raising exceptions in tf did not work in a saved model as you experienced the first time you used the combined policy and didn't pass the right model_selector name. I can remove this function and let _distribution just return something trivial or leave it empty (not sure if this would work), however, if I remember correctly every tf policy needs some type of distribution function.

File: compiler_opt/tools/combine_tfa_policies_lib.py (hunk 12 above)
Comment (tvmarino) [reply compiler_opt/tools/combine_tfa_policies_lib.py:32]: It's used to pass time_step_spec. I am keeping it.

File: compiler_opt/tools/combine_tfa_policies_lib.py (hunk 16 above)
Comment (tvmarino) [reply compiler_opt/tools/combine_tfa_policies_lib.py:124]: Added the issue and a TODO.

File: compiler_opt/tools/combine_tfa_policies.py (hunk 18 above)
Comment (tvmarino) [reply compiler_opt/tools/combine_tfa_policies.py:59]: Added the issue and a TODO.

File: compiler_opt/tools/combine_tfa_policies_lib.py (hunk 14 above)
Comment (mtrofin) [reply compiler_opt/tools/combine_tfa_policies_lib.py:84]: Ack on there needing to be a function.

Is it correct that the function isn't actually used during inference, and then it's just that we can't just throw (because of how tf evaluates it)... can we make it return something _invalid_ (and document what's going on, too)? Basically, we want to (1) be clear this is just an artifact of the framework (tf_agents), not actually needed, (2) make it fail-fast if our assumptions in (1) are invalidated in the future, which helps both maintaining the code, and a reader of the code.

File: compiler_opt/tools/combine_tfa_policies.py (hunk 11 above)
Comment (mtrofin) [reply compiler_opt/tools/combine_tfa_policies.py:58]: You added it when passing in the gin files as a command line param you mean?

Comment (tvmarino) [reply compiler_opt/tools/combine_tfa_policies.py:58]: Yes. I passed it as --gin_files='compiler_opt/rl/inlining/gin_configs/common.gin'.

File: compiler_opt/tools/combine_tfa_policies_lib.py (hunk 14 above)
Comment (tvmarino) [reply compiler_opt/tools/combine_tfa_policies_lib.py:84]: Set a default deterministic distribution as discussed.

**Review by mtrofin — APPROVED** [review]
